helix-view/src/stream.rs

```diff
@@ -0,0 +1,208 @@
+use std::io::Write;
+
+use anyhow::Error;
+use helix_core::{encoding, Rope, RopeBuilder};
+
+/// 8kB of buffer space for encoding and decoding `Rope`s.
+const BUF_SIZE: usize = 8192;
+
+#[derive(Default)]
+pub struct RopeWrite(RopeBuilder);
+impl RopeWrite {
+    pub fn finish(self) -> Rope {
+        self.0.finish()
+    }
+}
+
+impl Write for RopeWrite {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.0.append(std::str::from_utf8(buf).unwrap());
```

Comment:
`from_reader` guarantees that the input is always valid UTF8, but I wasn't sure if it would be appropriate to use `from_utf8_unchecked` here.

Reply:
Using `from_utf8_unchecked` would require you to mark this function unsafe as it is public and can potentially be called anywhere.

However we really want to avoid the ovheread of utf8 validation. Considering that you convert a `str` to bytes at the callsite I think it would be better to either:

* Use a custom `Write` trait that accepts `str` instead of `[u8]`
* use `std::fmt::Write`